File: snippets/design-patterns/behavioral/command.py
'''
--- Command Pattern ---
Use Case: Encapsulate requests or operations as objects to allow undoable actions or queuing.
Example: Task queues, undo/redo functionality, or macro recording.
'''

class Command:
    def execute(self):
        pass

    def undo(self):
        pass

class Light:
    def on(self):
        print("The light is ON")

    def off(self):
        print("The light is OFF")

class LightOnCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.on()

    def undo(self):
        self.light.off()

class LightOffCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.off()

    def undo(self):
        self.light.on()

# Example Usage
light = Light()
on_command = LightOnCommand(light)
off_command = LightOffCommand(light)

on_command.execute()  # Output: The light is ON
on_command.undo()     # Output: The light is OFF

File: snippets/design-patterns/behavioral/command.go
// --- Command Pattern ---
// Use Case: Encapsulate requests or operations as objects to allow undoable actions or queuing.
// Example: Task queues, undo/redo functionality, or macro recording.
// Description: The Command Pattern encapsulates a request as an object, allowing for
// parameterization of clients, logging of requests, and support for undoable operations.

package main

import "fmt"

type Command interface {
	Execute()
}

type Light struct{}

func (l *Light) On() {
	fmt.Println("Light is ON")
}

func (l *Light) Off() {
	fmt.Println("Light is OFF")
}

type LightOnCommand struct {
	light *Light
}

func (c *LightOnCommand) Execute() {
	c.light.On()
}

type LightOffCommand struct {
	light *Light
}

func (c *LightOffCommand) Execute() {
	c.light.Off()
}

type RemoteControl struct {
	command Command
}

func (r *RemoteControl) SetCommand(command Command) {
	r.command = command
}

func (r *RemoteControl) PressButton() {
	r.command.Execute()
}

func main() {
	light := &Light{}
	onCommand := &LightOnCommand{light: light}
	offCommand := &LightOffCommand{light: light}

	remote := &RemoteControl{}
	remote.SetCommand(onCommand)
	remote.PressButton() // Output: Light is ON

	remote.SetCommand(offCommand)
	remote.PressButton() // Output: Light is OFF
}
